### PR TITLE
[DRAFT] feat(v1): lazy loading custom selection set POC - implementation only

### PR DESCRIPTION
**Pull request by lawmicha**
## General Checklist

- [ ] Added new tests to cover change, if needed
- [ ] Build succeeds with all target using Swift Package Manager
- [ ] All unit tests pass
- [ ] All integration tests pass
- [ ] Security oriented best practices and standards are followed (e.g. using input sanitization, principle of least privilege, etc)
- [ ] Documentation update for the change if required
- [ ] PR title conforms to conventional commit style
- [ ] New or updated tests include `Given When Then` inline code documentation and are named accordingly `testThing_condition_expectation()`
- [ ] If breaking change, documentation/changelog update with migration instructions

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: AmplifyPlugins/Core/AWSPluginsCore/Model/GraphQLRequest/GraphQLRequest+AnyModelWithSync.swift [hunk 1]
```diff
@@ -169,15 +173,15 @@ extension GraphQLRequest: ModelSyncGraphQLRequestFactory {
                                     authType: AWSAuthorizationType? = nil) -> GraphQLRequest<MutationSyncResult> {
 
         var documentBuilder = ModelBasedGraphQLDocumentBuilder(modelSchema: modelSchema,
-                                                               operationType: .subscription)
+                                                               operationType: .subscription,
+                                                               primaryKeysOnly: true)
```
Comment (lawmicha) [inline AmplifyPlugins/Core/AWSPluginsCore/Model/GraphQLRequest/GraphQLRequest+AnyModelWithSync.swift:177]: The `subscription` API in GraphQLRequest+AnyModelWithSync will pass in `primaryKeysOnly` true initially. This is reverted back to false if it detects that the models do not have the lazy loading enabled. When lazy loading is enabled, then the selection set will be reduced to just the primary keys in the nested model, thus supporting mutation events sourced from other platforms. 

While testing Flutter with this backport, DataStore.start() will establish the subscriptions. The first step is to make sure the subscription's selection sets are reduced as expected.

File: AmplifyPlugins/API/AWSAPICategoryPlugin/Support/Decode/GraphQLResponseDecoder+DecodeData.swift [hunk 2]
```diff
@@ -66,4 +89,33 @@ extension GraphQLResponseDecoder {
...
+           request.responseType == MutationSync<AnyModel>.self,
+           case var .object(modelJSON) = graphQLData,
+           // No need to replace existing response payloads that have it already
+           modelJSON["__typename"] == nil {
+            modelJSON["__typename"] = .string(modelName)
+            return JSONValue.object(modelJSON)
+        } else {
+            return nil
+        }
+    }
```
Comment (lawmicha) [inline AmplifyPlugins/API/AWSAPICategoryPlugin/Support/Decode/GraphQLResponseDecoder+DecodeData.swift:120]: This is the second piece of logic that makes cross platform events successful, see in-line comment

**Comment by HuiSF** [comment]
From a few quick test runs, the custom selection set change eliminates the subscription GraphQL error described in this issue: https://github.com/aws-amplify/amplify-flutter/issues/663 . A create mutation originated from an Android device, can correctly trigger a onCreate event with correct payload delivered to the iOS App.

image

I'm looking forward to finalize this draft PR so we can resolve the linked issue for amplify-flutter developers and developers who develop Apps communicate between iOS and Android.

**Comment by lawmicha** [comment]
Closing for https://github.com/aws-amplify/amplify-swift/pull/2944

**Pull request closed without merging**